The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12674013805/job/35321747730) [(plain)](https://github.com/rust-lang/rust/commit/bed8b95b36c4d96c1ed05ac59665391414f4f35a/checks/35321747730/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 29.0s done
#21 DONE 35.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/slice-init.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll" "/checkout/tests/codegen/slice-init.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/slice-init.rs:77:12: error: CHECK: expected string not found in input
 // CHECK: llvm.memset.p0
           ^
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll:90:32: note: scanning from here
define void @u16_init_one_bytes(ptr dead_on_unwind noalias nocapture noundef writable sret([200 x i8]) align 2 dereferenceable(200) %_0) unnamed_addr #0 {
                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll:94:12: note: possible intended match here
 call void @llvm.lifetime.start.p0(i64 2, ptr %_2)

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll
Check file: /checkout/tests/codegen/slice-init.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'slice_init.587352d4ad61cb6b-cgu.0' 
            2: source_filename = "slice_init.587352d4ad61cb6b-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            5:  
            5:  
            6: %"core::option::Option<u8>" = type { i8, [1 x i8] } 
            7:  
            8: @0 = private unnamed_addr constant <{ [1 x i8], [1 x i8] }> <{ [1 x i8] zeroinitializer, [1 x i8] undef }>, align 1 
            9:  
           10: ; Function Attrs: nonlazybind uwtable 
           11: define void @zero_sized_elem() unnamed_addr #0 { 
           13:  %_1 = alloca [0 x i8], align 1 
           14: ; call slice_init::opaque 
           14: ; call slice_init::opaque 
           15:  call void @_ZN10slice_init6opaque17h9c01bc0e6568e7f7E(ptr noalias noundef nonnull readonly align 1 %_1) 
           16:  ret void 
           17: } 
           18:  
           19: ; Function Attrs: nonlazybind uwtable 
           20: define void @zero_len_array() unnamed_addr #0 { 
           22:  %_1 = alloca [0 x i8], align 4 
           23: ; call slice_init::opaque 
           23: ; call slice_init::opaque 
           24:  call void @_ZN10slice_init6opaque17h90ba94c9c1d53e83E(ptr noalias noundef nonnull readonly align 4 %_1) 
           25:  ret void 
           26: } 
           27:  
           28: ; Function Attrs: nonlazybind uwtable 
           29: define void @byte_array() unnamed_addr #0 { 
           30: start: 
           31:  %x = alloca [4 x i8], align 1 
           32:  call void @llvm.lifetime.start.p0(i64 4, ptr %x) 
           33:  call void @llvm.memset.p0.i64(ptr align 1 %x, i8 7, i64 4, i1 false) 
           34: ; call slice_init::opaque 
           35:  call void @_ZN10slice_init6opaque17h5a1c3b8d7de917daE(ptr noalias noundef readonly align 1 dereferenceable(4) %x) 
           36:  call void @llvm.lifetime.end.p0(i64 4, ptr %x) 
           37:  ret void 
           38: } 
           39:  
           40: ; Function Attrs: nonlazybind uwtable 
           41: define void @byte_enum_array() unnamed_addr #0 { 
           42: start: 
           43:  %x = alloca [4 x i8], align 1 
           44:  call void @llvm.lifetime.start.p0(i64 4, ptr %x) 
           45:  call void @llvm.memset.p0.i64(ptr align 1 %x, i8 1, i64 4, i1 false) 
           46: ; call slice_init::opaque 
           47:  call void @_ZN10slice_init6opaque17h4469190384ab35a6E(ptr noalias noundef readonly align 1 dereferenceable(4) %x) 
           48:  call void @llvm.lifetime.end.p0(i64 4, ptr %x) 
           49:  ret void 
           50: } 
           51:  
           52: ; Function Attrs: nonlazybind uwtable 
           53: define void @zeroed_integer_array() unnamed_addr #0 { 
           54: start: 
           55:  %x = alloca [16 x i8], align 4 
           56:  call void @llvm.lifetime.start.p0(i64 16, ptr %x) 
           57:  call void @llvm.memset.p0.i64(ptr align 4 %x, i8 0, i64 16, i1 false) 
           58: ; call slice_init::opaque 
           59:  call void @_ZN10slice_init6opaque17h587aa93d59c76bcaE(ptr noalias noundef readonly align 4 dereferenceable(16) %x) 
           60:  call void @llvm.lifetime.end.p0(i64 16, ptr %x) 
           61:  ret void 
           62: } 
           63:  
           64: ; Function Attrs: nonlazybind uwtable 
           65: define void @nonzero_integer_array() unnamed_addr #0 { 
           66: start: 
           67:  %x = alloca [16 x i8], align 4 
           68:  call void @llvm.lifetime.start.p0(i64 16, ptr %x) 
           69:  br label %repeat_loop_header 
           70:  
           71: repeat_loop_header: ; preds = %repeat_loop_body, %start 
           72:  %0 = phi i64 [ 0, %start ], [ %3, %repeat_loop_body ] 
           73:  %1 = icmp ult i64 %0, 4 
           74:  br i1 %1, label %repeat_loop_body, label %repeat_loop_next 
           75:  
           76: repeat_loop_body: ; preds = %repeat_loop_header 
           77:  %2 = getelementptr inbounds i32, ptr %x, i64 %0 
           78:  store i32 439041101, ptr %2, align 4 
           79:  %3 = add nuw i64 %0, 1 
           80:  br label %repeat_loop_header 
           81:  
           82: repeat_loop_next: ; preds = %repeat_loop_header 
           83: ; call slice_init::opaque 
           84:  call void @_ZN10slice_init6opaque17h587aa93d59c76bcaE(ptr noalias noundef readonly align 4 dereferenceable(16) %x) 
           85:  call void @llvm.lifetime.end.p0(i64 16, ptr %x) 
           86:  ret void 
           87: } 
           88:  
           89: ; Function Attrs: nonlazybind uwtable 
           90: define void @u16_init_one_bytes(ptr dead_on_unwind noalias nocapture noundef writable sret([200 x i8]) align 2 dereferenceable(200) %_0) unnamed_addr #0 { 
check:77'0                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:77'0     ~~~~~~~
           92:  %0 = alloca [2 x i8], align 2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           93:  %_2 = alloca [2 x i8], align 1 
           93:  %_2 = alloca [2 x i8], align 1 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           94:  call void @llvm.lifetime.start.p0(i64 2, ptr %_2) 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:77'1                ?                                        possible intended match
           95:  %1 = getelementptr inbounds i8, ptr %_2, i64 0 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           96:  store i8 1, ptr %1, align 1 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           97:  %2 = getelementptr inbounds i8, ptr %_2, i64 1 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           98:  store i8 1, ptr %2, align 1 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           99:  %_3 = load i16, ptr %_2, align 1, !noundef !3 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          100:  call void @llvm.lifetime.start.p0(i64 2, ptr %0) 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          101:  %3 = call i16 @llvm.bswap.i16(i16 %_3) 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          102:  store i16 %3, ptr %0, align 2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          103:  %_1 = load i16, ptr %0, align 2, !noundef !3 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          104:  call void @llvm.lifetime.end.p0(i64 2, ptr %0) 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          105:  call void @llvm.lifetime.end.p0(i64 2, ptr %_2) 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          106:  br label %repeat_loop_header 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          108: repeat_loop_header: ; preds = %repeat_loop_body, %start 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          109:  %4 = phi i64 [ 0, %start ], [ %7, %repeat_loop_body ] 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          110:  %5 = icmp ult i64 %4, 100 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
          111:  br i1 %5, label %repeat_loop_body, label %repeat_loop_next 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          113: repeat_loop_body: ; preds = %repeat_loop_header 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          114:  %6 = getelementptr inbounds i16, ptr %_0, i64 %4 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          115:  store i16 %_1, ptr %6, align 2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          116:  %7 = add nuw i64 %4, 1 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~
          117:  br label %repeat_loop_header 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          119: repeat_loop_next: ; preds = %repeat_loop_header 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          120:  ret void 
check:77'0     ~~~~~~~~~~
          121: } 
check:77'0     ~~
check:77'0     ~
check:77'0     ~
          123: ; Function Attrs: nonlazybind uwtable 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          124: define void @option_none_init(ptr dead_on_unwind noalias nocapture noundef writable sret([200 x i8]) align 1 dereferenceable(200) %_0) unnamed_addr #0 { 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          125: start: 
          126:  %0 = load i8, ptr @0, align 1, !range !4, !noundef !3 
          127:  %1 = trunc i8 %0 to i1 
          128:  %2 = load i8, ptr getelementptr inbounds (i8, ptr @0, i64 1), align 1 
          129:  br label %repeat_loop_header 
          130:  
          131: repeat_loop_header: ; preds = %repeat_loop_body, %start 
          132:  %3 = phi i64 [ 0, %start ], [ %8, %repeat_loop_body ] 
          133:  %4 = icmp ult i64 %3, 100 
          134:  br i1 %4, label %repeat_loop_body, label %repeat_loop_next 
          135:  
          136: repeat_loop_body: ; preds = %repeat_loop_header 
          137:  %5 = getelementptr inbounds %"core::option::Option<u8>", ptr %_0, i64 %3 
          138:  %6 = zext i1 %1 to i8 
          139:  store i8 %6, ptr %5, align 1 
          140:  %7 = getelementptr inbounds i8, ptr %5, i64 1 
          141:  store i8 %2, ptr %7, align 1 
          142:  %8 = add nuw i64 %3, 1 
          143:  br label %repeat_loop_header 
          144:  
          145: repeat_loop_next: ; preds = %repeat_loop_header 
          146:  ret void 
          147: } 
          149: ; slice_init::opaque 
          149: ; slice_init::opaque 
          150: ; Function Attrs: noinline nonlazybind uwtable 
          151: define void @_ZN10slice_init6opaque17h4469190384ab35a6E(ptr noalias noundef readonly align 1 dereferenceable(4) %_1) unnamed_addr #1 { 
          153:  ret void 
          154: } 
          155:  
          156: ; slice_init::opaque 
          156: ; slice_init::opaque 
          157: ; Function Attrs: noinline nonlazybind uwtable 
          158: define void @_ZN10slice_init6opaque17h587aa93d59c76bcaE(ptr noalias noundef readonly align 4 dereferenceable(16) %_1) unnamed_addr #1 { 
          160:  ret void 
          161: } 
          162:  
          163: ; slice_init::opaque 
          163: ; slice_init::opaque 
          164: ; Function Attrs: noinline nonlazybind uwtable 
          165: define void @_ZN10slice_init6opaque17h5a1c3b8d7de917daE(ptr noalias noundef readonly align 1 dereferenceable(4) %_1) unnamed_addr #1 { 
          167:  ret void 
          168: } 
          169:  
          170: ; slice_init::opaque 
          170: ; slice_init::opaque 
          171: ; Function Attrs: noinline nonlazybind uwtable 
          172: define void @_ZN10slice_init6opaque17h90ba94c9c1d53e83E(ptr noalias noundef nonnull readonly align 4 %_1) unnamed_addr #1 { 
          174:  ret void 
          175: } 
          176:  
          177: ; slice_init::opaque 
          177: ; slice_init::opaque 
          178: ; Function Attrs: noinline nonlazybind uwtable 
          179: define void @_ZN10slice_init6opaque17h9c01bc0e6568e7f7E(ptr noalias noundef nonnull readonly align 1 %_1) unnamed_addr #1 { 
          181:  ret void 
          182: } 
          183:  
          183:  
          184: ; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: write) 
          185: declare void @llvm.memset.p0.i64(ptr nocapture writeonly, i8, i64, i1 immarg) #2 
          186:  
          187: ; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none) 
          188: declare i16 @llvm.bswap.i16(i16) #3 
          189:  
          190: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
          191: declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #4 
          192:  
          193: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
          194: declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #4 
            .
            .
>>>>>>
------------------------------------------
```

</details>

--- a/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
@@ -93,22 +93,29 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     return;
                 }
 
+                let try_init_all_same = |bx: &mut Bx, v| {
+                    let start = dest.val.llval;
+                    let size = bx.const_usize(dest.layout.size.bytes());
+
+                    // Use llvm.memset.p0i8.* to initialize all zero arrays
+                    if bx.cx().const_to_opt_u128(v, false) == Some(0) {
+                        let fill = bx.cx().const_u8(0);
+                        bx.memset(start, fill, size, dest.val.align, MemFlags::empty());
+                        return true;
+                    }
+
+                    // Use llvm.memset.p0i8.* to initialize byte arrays
+                    let v = bx.from_immediate(v);
+                    if bx.cx().val_ty(v) == bx.cx().type_i8() {
+                        bx.memset(start, v, size, dest.val.align, MemFlags::empty());
+                        return true;
+                    }
+                    false
+                };
+
                 match cg_elem.val {
                     OperandValue::Immediate(v) => {
-                        let start = dest.val.llval;
-                        let size = bx.const_usize(dest.layout.size.bytes());
-
-                        // Use llvm.memset.p0i8.* to initialize all zero arrays
-                        if bx.cx().const_to_opt_u128(v, false) == Some(0) {
-                            let fill = bx.cx().const_u8(0);
-                            bx.memset(start, fill, size, dest.val.align, MemFlags::empty());
-                            return;
-                        }
-
-                        // Use llvm.memset.p0i8.* to initialize byte arrays
-                        let v = bx.from_immediate(v);
-                        if bx.cx().val_ty(v) == bx.cx().type_i8() {
-                            bx.memset(start, v, size, dest.val.align, MemFlags::empty());
+                        if try_init_all_same(bx, v) {
                             return;
                         }
                     }

--- a/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
@@ -93,23 +93,26 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     return;
                 }
 
-                if let OperandValue::Immediate(v) = cg_elem.val {
-                    let start = dest.val.llval;
-                    let size = bx.const_usize(dest.layout.size.bytes());
+                match cg_elem.val {
+                    OperandValue::Immediate(v) => {
+                        let start = dest.val.llval;
+                        let size = bx.const_usize(dest.layout.size.bytes());
 
-                    // Use llvm.memset.p0i8.* to initialize all zero arrays
-                    if bx.cx().const_to_opt_u128(v, false) == Some(0) {
-                        let fill = bx.cx().const_u8(0);
-                        bx.memset(start, fill, size, dest.val.align, MemFlags::empty());
-                        return;
-                    }
+                        // Use llvm.memset.p0i8.* to initialize all zero arrays
+                        if bx.cx().const_to_opt_u128(v, false) == Some(0) {
+                            let fill = bx.cx().const_u8(0);
+                            bx.memset(start, fill, size, dest.val.align, MemFlags::empty());
+                            return;
+                        }
 
-                    // Use llvm.memset.p0i8.* to initialize byte arrays
-                    let v = bx.from_immediate(v);
-                    if bx.cx().val_ty(v) == bx.cx().type_i8() {
-                        bx.memset(start, v, size, dest.val.align, MemFlags::empty());
-                        return;
+                        // Use llvm.memset.p0i8.* to initialize byte arrays
+                        let v = bx.from_immediate(v);
+                        if bx.cx().val_ty(v) == bx.cx().type_i8() {
+                            bx.memset(start, v, size, dest.val.align, MemFlags::empty());
+                            return;
+                        }
                     }
+                    _ => (),
                 }
 
                 let count = self

--- a/tests/codegen/slice-init.rs
+++ b/tests/codegen/slice-init.rs
@@ -63,6 +63,34 @@ pub fn nonzero_integer_array() {
     opaque(&x);
 }
 
+const N: usize = 100;
+
+// FIXME: The two bytes of the u16 are the same, so we should
+// just use memset, too.
+// CHECK-LABEL: @u16_init_one_bytes
+#[no_mangle]
+pub fn u16_init_one_bytes() -> [u16; N] {
+    // CHECK-NOT: select
+    // CHECK: br
+    // CHECK-NOT: switch
+    // CHECK: icmp
+    // CHECK-NOT: call void @llvm.memset.p0
+    [const { u16::from_be_bytes([1, 1]) }; N]
+}
+
+// FIXME: undef bytes can just be initialized with the same value as the
+// defined bytes, if the defines bytes are all the same.
+// CHECK-LABEL: @option_none_init
+#[no_mangle]
+pub fn option_none_init() -> [Option<u8>; N] {
+    // CHECK-NOT: select
+    // CHECK: br label %repeat_loop_header{{.*}}
+    // CHECK-NOT: switch
+    // CHECK: icmp
+    // CHECK-NOT: call void @llvm.memset.p0
+    [None; N]
+}
+
 // Use an opaque function to prevent rustc from removing useless drops.
 #[inline(never)]
 pub fn opaque(_: impl Sized) {}

